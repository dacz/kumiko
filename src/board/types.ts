export enum Direction {
  Up,
  Down,
}

export enum Filling {
  Empty,
  Star,
}

export interface Field {
  direction: Direction,
  filling: Filling,
}

export interface ColumnInput {
  jump?: number, // if not continued, +x is up, -x is down
  fields: Field[]
}

export interface Column {
  leftTopYCoord: number,
  leftBottomYCoord: number,
  rightTopYCoord: number,
  rightBottomYCoord: number,

  fields: Field[]
}

export type Columns = Column[]

export interface Board {
  columns: Columns,
  // topMax: () => number,
  // bottomMin: () => number,
  // width: () => number,
  // how the vertical lines are described?
  // how the down lines are calculated and described
  // how the up lines are calculated and described
}

