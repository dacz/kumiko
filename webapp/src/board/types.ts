export class GridDot {
  static readonly xSkew: number = Math.sqrt(0.75);

  x: number; // it is the index of the column (not real value, shorter than 1 - sqrt(0.5))
  y: number; // the real values

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  withinPaper(ps: PaperVirtualSize): boolean {
    return this.x >= 0 && this.x <= ps.maxX && this.y >= 0 && this.y <= ps.maxY;
  }

  downLeft(): GridDot { return new GridDot(this.x - 1, this.y - 0.5) }
  downRight(): GridDot { return new GridDot(this.x + 1, this.y - 0.5) }
  upLeft(): GridDot { return new GridDot(this.x - 1, this.y + 0.5) }
  upRight(): GridDot { return new GridDot(this.x + 1, this.y + 0.5) }
  up(): GridDot { return new GridDot(this.x, this.y + 1) }
  down(): GridDot { return new GridDot(this.x, this.y - 1) }
  clone(): GridDot { return new GridDot(this.x, this.y) }

  toRealDot(side: number): RealDot {
    return {
      x: this.x * side * GridDot.xSkew + side / 2,
      y: this.y * side + side / 2,
    }
  }
}

export interface RealDot {
  x: number,
  y: number,
}

export class Line {
  start: GridDot;
  end: GridDot;

  constructor(start: GridDot, end: GridDot) {
    this.start = start;
    this.end = end;
  }

  toRealLine(side: number): RealLine {
    return {
      start: this.start.toRealDot(side),
      end: this.end.toRealDot(side),
    }
  }
}

export interface RealLine {
  start: RealDot,
  end: RealDot,
}

// x is 0 and goes right
// y is 0 and goes up
export interface PaperVirtualSize {
  maxX: number,
  maxY: number,
}

export function serializePVS(pvs: PaperVirtualSize): string {
  return `PS:${pvs.maxX},${pvs.maxY}`;
}

export function deserializePVS(str: string): PaperVirtualSize {
  const [_, maxXs, maxYs] = str.match(/PS:(\d+),(\d+)$/) || [];
  const maxX = parseInt(maxXs, 10);
  const maxY = parseInt(maxYs, 10);
  if (isNaN(maxX) || isNaN(maxY)) throw new Error(`Invalid paper size: ${str}`);
  return { maxX, maxY };
}


