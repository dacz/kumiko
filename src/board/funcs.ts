import { Direction, Field, ColumnInput, Column, Board } from './types'

function verifyAlternatingFields(fields: Field[]): null | Error {
  if (fields.length == 0) {
    return new Error('no fields')
  }

  let prevDir: Direction | undefined
  let i = 0
  for (const field of fields) {
    i++
    if (prevDir === undefined) {
      prevDir = field.direction
      continue
    }

    if (prevDir === field.direction) {
      return new Error(`${i}`)
    }

    prevDir = field.direction
  }

  return null
}

// the first field of each column must be alternating if there is not a jump
function verifySufficientOverlapVerticalLines(cols: Column[]): null | Error {
  for (let i = 0; i < cols.length - 1; i++) {
    if (cols[i].rightTopYCoord < cols[i + 1].leftBottomYCoord) return new Error(`${i + 1}`)
    if (cols[i].rightBottomYCoord > cols[i + 1].leftTopYCoord) return new Error(`${i + 1}`)
  }

  return null
}

export function newBoard(input: ColumnInput[]): Board | Error {
  // verify fields alternating
  let i = 0
  for (const col of input) {
    i++
    const maybeErr = verifyAlternatingFields(col.fields)
    if (maybeErr instanceof Error) {
      return new Error(`fields must alternate - column ${i}: position: ${maybeErr.message}`)
    }
  }

  // calculate columns
  const [columns, n] = input.reduce((acc: [Column[], number], col) => {
    const nc = newColumn(acc[1], col)
    return [[...acc[0], nc], nc.rightTopYCoord]
  }, [[], 0] as [Column[], number])

  // check sufficient overlap of the vertial lines
  const maybeErr = verifySufficientOverlapVerticalLines(columns)
  if (maybeErr instanceof Error) {
    return new Error(`insufficient overlap between ${maybeErr.message}. and the next column`)
  }

  // check/identify points where only 2 lines are crossed (error, too - will be ugly)

  return {
    columns: columns,
  }
}

function newColumn(initPos: number, input: ColumnInput): Column {
  const leftTopYCoord = initPos + (input.jump || 0)
  const rightTopYCoord = leftTopYCoord + (
    input.fields[0].direction === Direction.Up ? 0.5 : -0.5
  )

  return {
    fields: input.fields,

    leftTopYCoord,
    leftBottomYCoord: leftTopYCoord - (
      input.fields[0].direction === Direction.Up
        ? Math.floor((input.fields.length) / 2)
        : Math.floor((input.fields.length + 1) / 2)
    ),
    rightTopYCoord,
    rightBottomYCoord: rightTopYCoord - (
      input.fields[0].direction === Direction.Up
        ? Math.floor((input.fields.length + 1) / 2)
        : Math.floor((input.fields.length) / 2)
    ),
  }
}

