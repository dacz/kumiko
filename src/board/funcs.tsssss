import { Direction, Field, ColumnInput, Column, Board, Line, GridDot } from './types'

function verifyAlternatingFields(fields: Field[]): null | Error {
  if (fields.length == 0) {
    return new Error('no fields')
  }

  let prevDir: Direction | undefined
  let i = 0
  for (const field of fields) {
    i++
    if (prevDir === undefined) {
      prevDir = field.direction
      continue
    }

    if (prevDir === field.direction) {
      return new Error(`${i}`)
    }

    prevDir = field.direction
  }

  return null
}

// the first field of each column must be alternating if there is not a jump
function verifySufficientOverlapVerticalLines(cols: Column[]): null | Error {
  for (let i = 0; i < cols.length - 1; i++) {
    if (cols[i].rightTopYCoord < cols[i + 1].leftBottomYCoord) return new Error(`${i + 1}`)
    if (cols[i].rightBottomYCoord > cols[i + 1].leftTopYCoord) return new Error(`${i + 1}`)
  }

  return null
}

// function ident2LinesUp(cols: Column[]) { }
// function ident2LinesDown(cols: Column[]) { }

// function makeGrid(cols: Column[]) { }

// TODO - zavest prazdny typ trojuhelniku - pro mezery

function leftLinesFromColumn(col: Column): Line[] { }
function rightLinesFromColumn(col: Column): Line[] { }
function deOverlapLinesOfSameDirection(lines: Line[]): Line[] { }
function upLinesFromColumn(col: Column): Line[] { }
function downLinesFromColumn(col: Column): Line[] { }
function linesOfPlane(plane: any, lines: Line[]): Line[] { } // plane is a function?, value?

// rewrite it to:
// 1. calculate the top left of the column (add to the column)
// 2. calculate vertical lines on both sides (for each column)
// 3. calculate the up/down lines of the fields (for each column)
// 4. calculate the biggest rectangle (for the whole board) (smallest and biggest x and y)



function makeLines(cols: Column[]): Line[] {
  const lines: Line[] = []
  for (let x = 0; x < cols.length; x++) {
    const col = cols[x]

    // -- vertical lines
    if (x == 0) {
      // left vertical line - is only one (which is generally not true)
      lines.push({
        start: { x, y: col.leftTopYCoord },
        end: { x, y: col.leftBottomYCoord },
      })
    }
    // right vertical line
    lines.push({
      start: { x, y: col.rightTopYCoord },
      end: { x, y: col.rightBottomYCoord },
    })

    for (let y = 0; y < col.fields.length; y++) {
      const field = col.fields[y]
      if (field.direction === Direction.Up) {
        // draw line from x, y to x + 0.5, y + 0.5
      } else {
        // draw line from x, y to x + 0.5, y - 0.5
      }
    }
  }

}

export function newBoard(input: ColumnInput[]): Board | Error {
  // verify fields alternating
  let i = 0
  for (const col of input) {
    i++
    const maybeErr = verifyAlternatingFields(col.fields)
    if (maybeErr instanceof Error) {
      return new Error(`fields must alternate - column ${i}: position: ${maybeErr.message}`)
    }
  }

  // calculate columns
  const [columns, _n] = input.reduce((acc: [Column[], number], col) => {
    const nc = newColumn(acc[1], col)
    return [[...acc[0], nc], nc.rightTopYCoord]
  }, [[], 0] as [Column[], number])

  // check sufficient overlap of the vertial lines
  const maybeErr = verifySufficientOverlapVerticalLines(columns)
  if (maybeErr instanceof Error) {
    return new Error(`insufficient overlap between ${maybeErr.message}. and the next column`)
  }

  // check/identify points where only 2 lines are crossed (error, too - will be ugly)

  return {
    columns: columns,
  }
}

// then calculate other things on the board 
//    * bounding box
//    * vertical coords []
//    * ups coords []
//    * downs coords []

function newColumn(initPos: number, input: ColumnInput): Column {
  const leftTopYCoord = initPos + (input.jump || 0)
  const rightTopYCoord = leftTopYCoord + (
    input.fields[0].direction === Direction.Up ? 0.5 : -0.5
  )

  return {
    fields: input.fields,

    leftTopYCoord,
    leftBottomYCoord: leftTopYCoord - (
      input.fields[0].direction === Direction.Up
        ? Math.floor((input.fields.length) / 2)
        : Math.floor((input.fields.length + 1) / 2)
    ),
    rightTopYCoord,
    rightBottomYCoord: rightTopYCoord - (
      input.fields[0].direction === Direction.Up
        ? Math.floor((input.fields.length + 1) / 2)
        : Math.floor((input.fields.length) / 2)
    ),
  }
}

