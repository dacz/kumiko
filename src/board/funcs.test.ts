import { Direction, Filling, Field, ColumnInput, Column, Board } from './types'
import { newBoard } from './funcs'
import { describe, it, expect } from 'vitest'

const colsInput: ColumnInput[] = [
  {
    fields: [
      { direction: Direction.Up, filling: Filling.Empty },
      { direction: Direction.Down, filling: Filling.Empty },
      { direction: Direction.Up, filling: Filling.Empty },
    ]
  },
  {
    // jump: -1,
    fields: [
      { direction: Direction.Down, filling: Filling.Empty },
      { direction: Direction.Up, filling: Filling.Empty },
      { direction: Direction.Down, filling: Filling.Empty },
    ]
  },
]

describe('newBoard', () => {
  it('ok', () => {
    const b = newBoard(colsInput)
    expect(b).not.toBeInstanceOf(Error)
    expect(b).toMatchSnapshot()
  })

  it('ok 2', () => {
    const colsInputCopy = structuredClone(colsInput)
    colsInputCopy[1] = structuredClone(colsInputCopy[0])
    const b = newBoard(colsInputCopy)
    console.log('b', b);
    expect(b).not.toBeInstanceOf(Error)
    expect(b).toMatchSnapshot()
  })

  it('ok, with jump', () => {
    const colsInputCopy = structuredClone(colsInput)
    colsInputCopy[1].jump = 1
    const b = newBoard(colsInputCopy)
    expect(b).not.toBeInstanceOf(Error)
    expect(b).toMatchSnapshot()
  })

  it('fail alternating fields', () => {
    const colsInputCopy = structuredClone(colsInput)
    colsInputCopy[1].fields[0].direction = Direction.Up
    const b = newBoard(colsInputCopy)
    expect(b).toBeInstanceOf(Error)
  })

  it('ok alternating columns if jump', () => {
    const colsInputCopy = structuredClone(colsInput)
    const secColumn = structuredClone(colsInputCopy[0])
    secColumn.jump = 1
    colsInputCopy[1] = secColumn
    const b = newBoard(colsInputCopy)
    expect(b).not.toBeInstanceOf(Error)
    expect(b).toMatchSnapshot()
  })

  it('ok 2 with jump, touching by corner (is ok)', () => {
    const colsInputCopy = structuredClone(colsInput)
    colsInputCopy[1] = structuredClone(colsInputCopy[0])
    colsInputCopy[1].jump = 1
    const b = newBoard(colsInputCopy)
    expect(b).not.toBeInstanceOf(Error)
    expect(b).toMatchSnapshot()
  })

  it('fail with 2 jumps, not touching', () => {
    const colsInputCopy = structuredClone(colsInput)
    colsInputCopy[1] = structuredClone(colsInputCopy[0])
    colsInputCopy[1].jump = 2
    const b = newBoard(colsInputCopy)
    expect(b).toMatchSnapshot()
    expect(b).toBeInstanceOf(Error)
  })
})
