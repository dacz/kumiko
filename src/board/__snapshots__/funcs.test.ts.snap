// Vitest Snapshot v1

exports[`newBoard > fail with 2 jumps, not touching 1`] = `[Error: insufficient overlap between 1. and the next column]`;

exports[`newBoard > ok 1`] = `
{
  "columns": [
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1,
      "leftTopYCoord": 0,
      "rightBottomYCoord": -1.5,
      "rightTopYCoord": 0.5,
    },
    {
      "fields": [
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1.5,
      "leftTopYCoord": 0.5,
      "rightBottomYCoord": -1,
      "rightTopYCoord": 0,
    },
  ],
}
`;

exports[`newBoard > ok 2 1`] = `
{
  "columns": [
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1,
      "leftTopYCoord": 0,
      "rightBottomYCoord": -1.5,
      "rightTopYCoord": 0.5,
    },
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -0.5,
      "leftTopYCoord": 0.5,
      "rightBottomYCoord": -1,
      "rightTopYCoord": 1,
    },
  ],
}
`;

exports[`newBoard > ok 2 with jump, touching by corner (is ok) 1`] = `
{
  "columns": [
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1,
      "leftTopYCoord": 0,
      "rightBottomYCoord": -1.5,
      "rightTopYCoord": 0.5,
    },
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": 0.5,
      "leftTopYCoord": 1.5,
      "rightBottomYCoord": 0,
      "rightTopYCoord": 2,
    },
  ],
}
`;

exports[`newBoard > ok alternating columns if jump 1`] = `
{
  "columns": [
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1,
      "leftTopYCoord": 0,
      "rightBottomYCoord": -1.5,
      "rightTopYCoord": 0.5,
    },
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": 0.5,
      "leftTopYCoord": 1.5,
      "rightBottomYCoord": 0,
      "rightTopYCoord": 2,
    },
  ],
}
`;

exports[`newBoard > ok, with jump 1`] = `
{
  "columns": [
    {
      "fields": [
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -1,
      "leftTopYCoord": 0,
      "rightBottomYCoord": -1.5,
      "rightTopYCoord": 0.5,
    },
    {
      "fields": [
        {
          "direction": 1,
          "filling": 0,
        },
        {
          "direction": 0,
          "filling": 0,
        },
        {
          "direction": 1,
          "filling": 0,
        },
      ],
      "leftBottomYCoord": -0.5,
      "leftTopYCoord": 1.5,
      "rightBottomYCoord": 0,
      "rightTopYCoord": 1,
    },
  ],
}
`;
